README.md
# forkinfo

Github will show you a list of forks for any repository via its Insights page, but to find out any other
useful information such as when the fork was last updated, how many commits differ etc., you're
forced to open each and every fork in a new tab... Often to find that most haven't changed at all!

Forkinfo aims to fix that by providing basic information about all forks with one simple command.

## Installation
```
go get github.com/dshoreman/forkinfo
```

Alternatively, the binary can be installed manually:
```
git clone -b master https://github.com/dshoreman/forkinfo.git
make
sudo make install
```

### Development

To build Forkinfo without installing globally, clone this repo and run `go build` in the project directory.
This will compile the application to `forkinfo` which can be run as `./forkinfo <user>/<repo>`.

Alternatively, you can use `go run forkinfo.go` which will compile into `/tmp` and run it automatically.

## Usage

Forkinfo takes a single argument, and that's the `user/repo` string. For example, to check this repo:

```
$ forkinfo dshoreman/forkinfo
```

When you run Forkinfo you will be prompted to enter a Personal Access Token to avoid hitting rate
limits on the Github API. This is optional, but will be saved to a config file once you paste in your
token and hit enter. To skip authentication, run Forkinfo with the `--no-token` flag. Note that "guest"
API access is limited to 60 requests per hour, which may be too low for some repositories.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).


## [Unreleased]


## [0.2.1] - 2019-06-18
### Fixed
* Changelog diff links were pointing to Nextshot


## [0.2.0] - 2019-06-18
### Added
* Support for API authentication
* Summary of forks (total, unchanged, outdated or broken)
* Number of ahead/behind/total commits shown in fork stats
* Option (`--show-commits`) for listing commits of ahead or diverged forks
* Ability to hide old (`--hide-old`) and duplicate (`--hide-same`) forks


## [0.1.0] - 2019-05-20
### Added
* Basic repository statistics
* Listing of repository's forks


[Unreleased]: https://github.com/dshoreman/forkinfo/compare/v0.2.0...develop
[0.2.0]: https://github.com/dshoreman/forkinfo/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/dshoreman/forkinfo/releases/tag/v0.1.0


config.go
package main

import (
    "encoding/json"
    "fmt"
    "io/ioutil"
    "os"
    "strings"
)

const configFile = "config.json"
const configPath = ".config/forkinfo"

var config Config

type Config struct {
    AccessToken string `json:"access_token"`
}

func configFullPath() string {
    return strings.Join([] string {
        os.Getenv("HOME"),
        configPath,
        configFile,
    }, "/")
}

func loadConfig() {
    if data, err := ioutil.ReadFile(configFullPath()); err == nil {
        json.Unmarshal(data, &config)
    } else if !os.IsNotExist(err) {
        abortOnError(err)
    }
}

func writeConfig() {
    fmt.Println("Saving config to ", configFullPath(), "...")
    configString, _ := json.MarshalIndent(config, "", "  ")

    os.MkdirAll(strings.Join([] string {os.Getenv("HOME"), configPath}, "/"), 0700)

    if err := ioutil.WriteFile(configFullPath(), append(configString, '\n'), 0644); err != nil {
        fmt.Println("Failed saving config")
        fmt.Println(err)
    }
    loadConfig()
}


forkinfo.go
package main

import (
    "bufio"
    "context"
    "fmt"
    "net/http"
    "time"
    "os"
    "strconv"
    "strings"

    flag "github.com/ogier/pflag"
    "github.com/google/go-github/github"
    "golang.org/x/oauth2"
)

const version = "0.2.0"

var (
    authClient *http.Client
    client *github.Client
    ctx = context.Background()
    hideDupes bool
    hideOld bool
    saveConfig bool
    showCommits bool
    skipAuth bool
)

func setupAPI() {
    if saveConfig {
        writeConfig()
    }
    if !skipAuth {
        authClient = oauth2.NewClient(ctx, oauth2.StaticTokenSource(
            &oauth2.Token{AccessToken: config.AccessToken},
        ))
    }
    client = github.NewClient(authClient)
}

func promptForToken() {
    fmt.Println("The Github API limits Unauthenticated access to 60 requests per")
    fmt.Println("hour. To raise these limits, create a Personal Access Token at")
    fmt.Println("https://github.com/settings/tokens/new?description=Forkinfo.")
    fmt.Println("Leaves scopes unchecked - Forkinfo requires no special access.")
    fmt.Println("To run without authentication, use forkinfo with `--no-token`.")
    fmt.Println()

    reader := bufio.NewReader(os.Stdin)
    for config.AccessToken == "" {
        fmt.Println("Paste your personal access token:")
        token, _ := reader.ReadString('\n')
        config.AccessToken = strings.Trim(token, " \r\n\t")
    }
}

func fetchRepository(username, repository string) (repo *github.Repository) {
    repo, _, err := client.Repositories.Get(ctx, username, repository)
    abortOnError(err)
    return
}

func fetchRepositoryForks(repo *github.Repository) (forks []*github.Repository) {
    opts := github.RepositoryListForksOptions{
        ListOptions: github.ListOptions{PerPage: repo.GetForksCount()},
    }
    forks, _, err := client.Repositories.ListForks(ctx, *repo.Owner.Login, *repo.Name, &opts)
    abortOnError(err)
    return
}

func listDifferingCommits(cmp *github.CommitsComparison) {
    for _, commit := range cmp.Commits {
        c := commit.GetCommit()
        fmt.Printf(
            "%s\tAuthored by %s at %s\n%s\n\n",
            commit.GetSHA()[:8],
            c.GetAuthor().GetName(),
            c.GetAuthor().GetDate().Format("15:04 on _2 Jan 2006"),
            c.GetMessage(),
        )
    }
    fmt.Printf("Diff: %s\n", cmp.GetHTMLURL())
}

func printRepoStats(repo *github.Repository, format string) {
    var datefmt string
    output := "Watchers: %d\tStars: %d\tForks: %d"

    if format == "short" {
        datefmt = "02/01/2006 15:04:05"
        output += "\tIssues/PRs: %d\tLast push: %s\n\n"
    } else {
        datefmt = string(time.RFC1123)
        output += "\n\nThis repository has %d open issues and PRs\nMost recent push: %s\n\n"
    }

    fmt.Printf(
        output,
        repo.GetSubscribersCount(),
        repo.GetStargazersCount(),
        repo.GetForksCount(),
        repo.GetOpenIssuesCount(),
        repo.PushedAt.Format(datefmt),
    )
}

func rowNum(row, total int) string {
    return fmt.Sprintf("[%*d/%d]", len(strconv.Itoa(total)), row, total)
}

func main() {
    args := strings.Split(flag.Arg(0), "/")
    username, repository := args[0], args[1]

    if config.AccessToken == "" && !skipAuth  {
        promptForToken()
        saveConfig = true
    }
    setupAPI()

    fmt.Println("Fetching repository...")
    repo := fetchRepository(username, repository)

    fmt.Printf("\n%s: %s\n - %s\n\n", *repo.Name, *repo.Description, *repo.HTMLURL)
    printRepoStats(repo, "long")

    if repo.GetForksCount() == 0 {
        return
    }

    fmt.Printf("Listing forks of %s...\n\n", *repo.FullName)
    forks := fetchRepositoryForks(repo)
    numBad := 0
    numOld := 0
    numDupes := 0
    numForks := len(forks)

    for i, fork := range forks {
        if fork.PushedAt.Equal(*repo.PushedAt) {
            numDupes++
            if hideDupes {
                continue
            }
        }

        base := "master"
        head := fmt.Sprintf("%s:master", *fork.Owner.Login)
        comp, _, err := client.Repositories.CompareCommits(ctx, username, *repo.Name, base, head)
        if err != nil {
            numBad++
            continue
        }
        if comp.GetStatus() == "identical" {
            numDupes++
            if hideDupes {
                continue
            }
        }
        if comp.GetStatus() == "behind" {
            numOld++
            if hideOld {
                continue
            }
        }

        fmt.Printf("%s %s\n", rowNum(i+1, numForks), *fork.FullName)
        fmt.Printf(
            "Status: %s (%d commits - %d ahead, %d behind)\n",
            comp.GetStatus(),
            comp.GetTotalCommits(),
            comp.GetAheadBy(),
            comp.GetBehindBy(),
        )
        printRepoStats(fork, "short")

        if showCommits && (comp.GetStatus() == "diverged" || comp.GetStatus() == "ahead") {
            listDifferingCommits(comp)
        }
        fmt.Println("------------------------------------------------------------------------")
    }

    fmt.Println("Fork listing complete.")
    fmt.Printf("Of the %d forks found, there were:\n  %d unchanged mirrors\n", numForks, numDupes)
    fmt.Printf("  %d outdated mirrors\n  %d broken forks\n", numOld, numBad)
}

func init() {
    flag.Usage = func() {
        fmt.Fprintf(os.Stderr, "Usage: %s [options] <user>/<repository>\n", os.Args[0])
        fmt.Fprintf(os.Stderr, "\nOptions:\n\n")
        flag.PrintDefaults()
    }

    flag.BoolVar(&hideDupes, "hide-same", false, "Hide unchanged mirrors")
    flag.BoolVar(&hideOld, "hide-old", false, "Hide outdated mirrors")
    flag.BoolVar(&showCommits, "show-commits", false, "Show commits for non-mirror forks")
    token := flag.StringP("token", "t", "", "Set the Personal Access Token for API authentication.")
    flag.BoolVarP(&skipAuth, "no-token", "T", false, "Use the Github API without authentication.")
    showVersionInfo := flag.BoolP("version", "V", false, "Print version info and quit.")
    flag.Parse()
    loadConfig()

    if *showVersionInfo {
        fmt.Println("Forkinfo " + version)
        os.Exit(0)
    }
    validate(*token)
    if *token != "" {
        config.AccessToken = *token
    } else if skipAuth {
        fmt.Println("Running without authentication.")
    }
}

func validate(token string) {
    if skipAuth && token != "" {
        abort("Cannot skip authentication while also passing an access token.")
    }
    if len(flag.Args()) == 0 {
        abort("Not enough arguments supplied.")
    }
    if len(flag.Args()) > 1 {
        abort("Too many arguments supplied.")
    }
    if ! strings.Contains(flag.Arg(0), "/") {
        abort("Argument is not a valid user/repo string.")
    }
}

func abort(msg string) {
    fmt.Printf("[ERROR] %s\n\n", msg)
    flag.Usage()
    os.Exit(1)
}

func abortOnError(err error) {
    if err != nil {
        fmt.Printf("[ERROR] %v\n", err)
        os.Exit(1)
    }
}
